File: db/unixfs/rpc/client/fs-handle.go
package unixfs_rpc_client

import (
	"context"

	"github.com/s4wave/spacewave/db/unixfs"
	unixfs_rpc "github.com/s4wave/spacewave/db/unixfs/rpc"
)

// BuildFSHandle constructs a root fs handle from a FSCursor service.
//
// rctx is used as the root context for the rpc client
// rctx must remain not-canceled during the duration of the lifetime of the FSHandle.
func BuildFSHandle(rctx context.Context, fsCursorSvcClient unixfs_rpc.SRPCFSCursorServiceClient) (*unixfs.FSHandle, error) {
	fsCursor := NewFSCursor(rctx, fsCursorSvcClient)
	fs, err := unixfs.NewFSHandle(fsCursor)
	if err != nil {
		fsCursor.Release()
		return nil, err
	}
	return fs, nil
}

// NewFSHandleBuilder constructs a new FSHandleBuilder with a FSCursorServiceClient.
func NewFSHandleBuilder(fsCursorSvcClient unixfs_rpc.SRPCFSCursorServiceClient) unixfs.FSHandleBuilder {
	return func(ctx context.Context, released func()) (*unixfs.FSHandle, func(), error) {
		fs, err := BuildFSHandle(ctx, fsCursorSvcClient)
		if err != nil {
			return nil, nil, err
		}
		fs.AddReleaseCallback(released)
		return fs, fs.Release, nil
	}
}

File: db/unixfs/rpc/client/fs-handle.ts
import { FSHandle } from '../../fs-handle.js'
import type { FSCursorService } from '../rpc_srpc.pb.js'
import { RpcFSCursor } from './fs-cursor.js'

// FSHandleBuilder constructs a new FSHandle from a signal and release callback.
export type FSHandleBuilder = (
  signal: AbortSignal,
  released: () => void,
) => Promise<{ handle: FSHandle; release: () => void }>

// buildFSHandle constructs a root FSHandle from a FSCursorService client.
// The signal must remain active for the duration of the FSHandle's lifetime.
export function buildFSHandle(
  svcClient: FSCursorService,
  signal?: AbortSignal,
): Promise<FSHandle> {
  const cursor = new RpcFSCursor(
    signal ?? new AbortController().signal,
    svcClient,
  )
  try {
    return Promise.resolve(FSHandle.create(cursor))
  } catch (e) {
    cursor.release()
    throw e
  }
}

// newFSHandleBuilder constructs a new FSHandleBuilder for a FSCursorService client.
export function newFSHandleBuilder(
  svcClient: FSCursorService,
): FSHandleBuilder {
  return async (signal: AbortSignal, released: () => void) => {
    const handle = await buildFSHandle(svcClient, signal)
    handle.addReleaseCallback(released)
    return { handle, release: () => handle.release() }
  }
}
